doc(tuto): DPF collections tutorial (#2742)
pyansys/pydpf-core: doc/source/user_guide/tutorials/data_structures/collections.rst

.. _ref_tutorials_collections:

===============
DPF Collections
===============

.. include:: ../../links_and_refs.rst

> This tutorial demonstrates how to create and work with some DPF collections: FieldsContainer, MeshesContainer and ScopingsContainer.

Reviewer: ```suggestion
This tutorial demonstrates how to create and work with some DPF collections: |FieldsContainer|, |MeshesContainer| and |ScopingsContainer|.
```

> DPF collections are homogeneous groups of labeled raw data storage structures that allow you to organize and manipulate related data efficiently. Collections are essential for handling multiple time steps, frequency sets, or other labeled datasets in your analysis workflows.

Reviewer: What do you mean by "homogeneous groups"?
I personally don't like to use the term "Labeled " here because the concept of label in DFP has not been explained yet. "Collections gather  raw data storage structures grouped by a category. In dpf these categories are labels ..." Something like this.

:jupyter-download-script:`Download tutorial as Python script<collections>`
:jupyter-download-notebook:`Download tutorial as Jupyter notebook<collections>`

Introduction to Collections
---------------------------

Collections in DPF serve as containers that group related objects with labels. The main collection types are:

- |FieldsContainer|: A collection of |Field| objects, typically representing results over multiple time steps or frequency sets

> - |MeshesContainer|: A collection of |MeshedRegion| objects for different configurations or time steps  

Reviewer: The term "configurations" her can be misunderstood as it is also a term in the DPF vocabulary.

- |ScopingsContainer|: A collection of |Scoping| objects for organizing entity selections

Each collection provides methods to:

- Add, retrieve, and iterate over contained objects

> - Access objects by label (time, frequency, set ID, etc.)

Reviewer 2: ```suggestion
- Access objects by label (time, frequency, set ID, and so on)
```

- Perform operations across all contained objects

Collections are widely used in DPF workflows to provide vectorized data to operators,
allowing you to process the data in bulk or to process it in parallel whenever possible.

Set up the Analysis
-------------------

First, import the required modules and load a transient analysis result file that contains multiple time steps.

.. jupyter-execute::

    # Import the ansys.dpf.core module
    import ansys.dpf.core as dpf
    
    # Import the examples module 
    from ansys.dpf.core import examples
    
    # Load a transient analysis with multiple time steps
    result_file_path = examples.find_msup_transient()
    
    # Create a DataSources object
    data_sources = dpf.DataSources(result_path=result_file_path)
    
    # Create a Model from the data sources
    model = dpf.Model(data_sources=data_sources)
    
    # Display basic model information
    print(model)

> Working with FieldsContainer

Reviewer: As we repeat this section for the different collection types and to avoid a giant tutorial page, I would have them in tabs.

A |FieldsContainer| is the most commonly used collection in DPF. It stores multiple |Field| objects, each associated with a label such as time step or frequency.

Extract Results into a FieldsContainer
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Let's extract displacement results for all time steps, which will automatically create a |FieldsContainer|.

.. jupyter-execute::

    # Get displacement results for all time steps
    displacement_fc = model.results.displacement.eval()
    
    # Display FieldsContainer information
    print(displacement_fc)

Access Individual Fields in the Container
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

You can access individual fields by their label or index.

.. jupyter-execute::

    # Access field by index (first time step)
    first_field = displacement_fc[0]
    print(f"First field info:")
    print(f"  Location: {first_field.location}")
    print(f"  Number of entities: {first_field.scoping.size}")

>     print(f"  Components: {first_field.component_count}")

Author: ```suggestion
    print(first_field)
```

            field.scoping.ids = [1, 2, 3, 4, 5]
            field.data = [float(time_step * zone * i) for i in range(1, 6)]
            
            # Add field to container with labels
            custom_fc.add_field({"time": time_step, "zone": zone}, field)
    
    # Display the custom FieldsContainer
    print(custom_fc)

Working with ScopingsContainer
------------------------------

A |ScopingsContainer| holds multiple |Scoping| objects, which define sets of entity IDs (nodes, elements, etc.).

Create and Populate a ScopingsContainer
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Create different node selections and organize them in a |ScopingsContainer|.

.. jupyter-execute::

    # Get the mesh from our model
    mesh = model.metadata.meshed_region
    
    # Create a ScopingsContainer
    scopings_container = dpf.ScopingsContainer()
    # Set labels for different selections
    scopings_container.labels = ["selection_type"]
    # Selection 1: First 10 nodes
    first_nodes = dpf.Scoping(location=dpf.locations.nodal)
    first_nodes.ids = list(range(1, 11))
    scopings_container.add_scoping(label_space={"selection_type": 0}, scoping=first_nodes)

>     # Selection 2: Every 10th node (sample)

Reviewer: I would put more space between the blocks so it can be more visual.

    subset_mesh = mesh_extract_op.eval()

    # Add subset mesh to container
    meshes_container.add_mesh({"variation": 1}, subset_mesh)

    # Display MeshesContainer information
    print(meshes_container)

Collection Operations and Iteration
------------------------------------

Collections support various operations for data manipulation and analysis.

Iterate Through Collections
^^^^^^^^^^^^^^^^^^^^^^^^^^^

You can iterate through collections using different methods.

.. jupyter-execute::

    # Iterate through FieldsContainer by index
    print("Iterating through displacement fields by index:")
    for i in range(min(3, len(displacement_fc))):  # Show first 3 fields
        field = displacement_fc[i]
        label_space = displacement_fc.get_label_space(i)
        max_value = field.data.max()
        print(f"  Field {i}: {label_space}, max value: {max_value:.6f}")
    
    print("\nIterating through ScopingsContainer:")
    for i, scoping in enumerate(scopings_container):
        label_space = scopings_container.get_label_space(i)
        print(f"  Scoping {i}: {label_space}, size: {scoping.size}")

Filter and Select from Collections  
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> You can filter collections based on labels or criteria.

Reviewer: Which criteria?

        label_space = displacement_fc.get_label_space(i)
        max_value = field.data.max()
        print(f"  Field {i}: {label_space}, max value: {max_value:.6f}")
    
    print("\nIterating through ScopingsContainer:")
    for i, scoping in enumerate(scopings_container):
        label_space = scopings_container.get_label_space(i)
        print(f"  Scoping {i}: {label_space}, size: {scoping.size}")

Filter and Select from Collections  
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

You can filter collections based on labels or criteria.

.. jupyter-execute::

    # Get specific fields from FieldsContainer by label criteria
    if len(displacement_fc) >= 2:
        # Get the second time step
        time_sets = list(displacement_fc.get_label_space(0).keys())
        if len(time_sets) > 1:
            specific_field = displacement_fc.get_field({"time": time_sets[1]})
            print(f"Retrieved field for time {time_sets[1]}:")
            print(f"  Components: {specific_field.component_count}")

>             print(f"  Location: {specific_field.location}")

Author: ```suggestion
    # Get all fields of ``custom_fc`` where ``zone=1``
    zone_1_fields = custom_fc.get_fields({"zone": 1})
    print(zone_1_fields)
```

    
    # Create a DataSources object
    data_sources = dpf.DataSources(result_path=result_file_path)
    
    # Create a Model from the data sources
    model = dpf.Model(data_sources=data_sources)
    
    # Display basic model information
    print(f"Number of time steps: {len(model.metadata.time_freq_support.time_frequencies)}")
    print(f"Available results: {model.metadata.result_info.available_results}")

Working with FieldsContainer
-----------------------------

A |FieldsContainer| is the most commonly used collection in DPF. It stores multiple |Field| objects, each associated with a label such as time step or frequency.

Extract Results into a FieldsContainer
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Let's extract displacement results for all time steps, which will automatically create a |FieldsContainer|.

.. jupyter-execute::

    # Get displacement results for all time steps
    displacement_fc = model.results.displacement.eval()
    
    # Display FieldsContainer information
    print(f"Type: {type(displacement_fc)}")
    print(f"Number of fields: {len(displacement_fc)}")
    print(f"Labels: {displacement_fc.labels}")

>     print(f"Available time sets: {list(displacement_fc.get_label_space(0).keys())}")

Author: ```suggestion
    print(displacement_fc)
```

        # Get the second time step
        time_sets = list(displacement_fc.get_label_space(0).keys())
        if len(time_sets) > 1:
            specific_field = displacement_fc.get_field({"time": time_sets[1]})
            print(f"Retrieved field for time {time_sets[1]}:")
            print(f"  Components: {specific_field.component_count}")
            print(f"  Location: {specific_field.location}")
    
    # Get scoping by selection criteria
    first_ten_scoping = scopings_container.get_scoping({"selection_type": 0})
    print(f"\nRetrieved 'first_ten' scoping:")
    print(f"  Size: {first_ten_scoping.size}")
    print(f"  First 5 IDs: {first_ten_scoping.ids[:5]}")

Collection Summary and Best Practices
--------------------------------------

Let's summarize the key concepts and best practices for working with DPF collections.

.. jupyter-execute::

    print("DPF Collections Summary:")
    print("=" * 50)
    
    print(f"\n1. FieldsContainer:")
    print(f"   - Purpose: Store multiple Field objects with labels")
    print(f"   - Common use: Results over time steps, frequencies, or load cases")
    print(f"   - Example size: {len(displacement_fc)} fields")

>     print(f"   - Labels: {displacement_fc.labels()}")

Author: ```suggestion
```

.. _ref_tutorials_collections:

===============
DPF Collections

> ===============

Reviewer 3: ```suggestion
###############
DPF Collections
###############
```

Suggestion should be accepted if you decide that the new "What You'll Learn" section should be kept

The :py:meth:`ansys.dpf.core.collection.Collection.collection_factory` method allows you to create a collection class for any supported DPF type at runtime.
This is useful when you want to group and manage objects that are not covered by the built-in collection types (such as FieldsContainer, MeshesContainer, or ScopingsContainer).

For example, you can create a collection for :class:`ansys.dpf.core.DataSources` objects:

.. jupyter-execute::

    from ansys.dpf.core import DataSources
    from ansys.dpf.core import examples
    from ansys.dpf.core.collection import Collection

    # Create a collection class for DataSources
    DataSourcesCollection = Collection.collection_factory(DataSources)
    ds_collection = DataSourcesCollection()
    ds_collection.labels = ["case"]

    # Add DataSources objects to the collection
    ds1 = DataSources("path/to/first/result/file.rst")
    ds2 = DataSources("path/to/second/result/file.rst")
    ds_collection.add_entry({"case": 0}, ds1)
    ds_collection.add_entry({"case": 1}, ds2)

    # Show the collection

>     print(ds_collection)

Reviewer 3: > The documentation for this pull request will be available at https://dpf.docs.pyansys.com/pull/2742. Please allow some time for the documentation to be deployed.

It is ok.

> This approach allows you to leverage the powerful labeling and grouping features of DPF collections for any supported DPF object type,

Reviewer 3: ```suggestion
This approach allows you to leverage the powerful labeling and grouping features of DPF collections for any supported DPF type,
```

To be consistent with rest of the tutorial ("type" or "object" have been used interchangeably but never together since they mean the same thing in this context).

which allows you to organize and keep track of data.
Collections are essential for handling multiple time steps, frequency sets, or other labeled datasets in your analysis workflows.

:jupyter-download-script:`Download tutorial as Python script<collections>`
:jupyter-download-notebook:`Download tutorial as Jupyter notebook<collections>`

What You'll Learn
=================

This tutorial covers the following topics:

- Working with |FieldsContainer|: Extract results across time steps, access individual fields, and create custom containers with multiple labels
- Working with |ScopingsContainer|: Create and manage selections, and use them with operators for targeted result extraction
- Working with |MeshesContainer|: Store and organize multiple mesh variations or time-dependent meshes
- Collection operations: Iterate through collections, filter by labels, and access metadata
- Advanced usage: Learn about other built-in collection types and create custom collections using the collection factory

By the end of this tutorial, you'll have a basic understanding of how to effectively organize and manipulate DPF data using collections in your analysis workflows.

> 

Reviewer 3: ```suggestion
```

I think the point of this subsection was marking the end of introduction/preliminary explanations and serving as the beginning of the core tutorial.

Otherwise the rendering appears this way:

<img width="1640" height="764" alt="Image" src="https://github.com/user-attachments/assets/16ddd45d-42fd-494f-bd16-09d32abfb3de" />

- a collection of fields (a |FieldsContainer|) across frequency and stage uses the labels **frequency** and **stage**
  with each field associated to a **frequency** integer value (the frequency ID) and a **stage** integer value (the stage ID).
  The **LabelSpace** ``{"frequency": 2, "stage": 1}`` would uniquely identify the field for frequency ID 2 at stage ID 1.
- a collection of meshes (a |MeshesContainer|) across different parts, further split by element type, uses the labels **part** and **element_type**
  with each mesh associated to a **part** integer value (the part ID) and an **element_type** integer value (the element type code).
  The **LabelSpace** ``{"part": 2, "element_type": 1}`` would uniquely identify the mesh for part 2 with element type code 1.
- a collection of scopings (a |ScopingsContainer|) across different fluid zones for a transient analysis uses the labels **zone** and **time**
  with each scoping associated to a **zone** integer value (the zone ID) and a **time** integer value (the time step ID).
  The **LabelSpace** ``{"zone": 1}`` would identify the collection of scopings for zone 1 only, for all time steps.

> 

Reviewer 3: ```suggestion
What You'll Learn
=================

This tutorial covers the following topics:

- Working with |FieldsContainer|: Extract results across time steps, access individual fields, and create custom containers with multiple labels
- Working with |ScopingsContainer|: Create and manage selections, and use them with operators for targeted result extraction
- Working with |MeshesContainer|: Store and organize multiple mesh variations or time-dependent meshes
- Collection operations: Iterate through collections, filter by labels, and access metadata
- Advanced usage: Learn about other built-in collection types and create custom collections using the collection factory

By the end of this tutorial, you'll have a basic understanding of how to effectively organize and manipulate DPF data using collections in your analysis workflows.

```
